Added private endpoint option to GKE setup

Spazzy757 opened the pull request:
## Problem
By default GKE assigns an IP address to every node that it spins up, this in itself is not a problem however unless requested a Google Cloud Project has Quotas on the amount of Public IP's a project can have (I believe the default is 8). Therefore if you are not careful your cluster can not scale or do rolling updates do to insufficient IP's 

## Changelog

- Adds the ability to have private endpoints in the Google Kubernetes Cluster

pst commented on google/cluster/configuration.tf:
```diff
@@ -56,4 +56,9 @@ locals {
+
+  enable_private_nodes    = lookup(local.cfg, "enable_private_nodes", true)
+  enable_private_endpoint = lookup(local.cfg, "enable_private_endpoint", false)
+  master_cidr_block       = lookup(local.cfg, "master_cidr_block", "10.254.0.0/28")
```
Why this default? Is there a default on Google's side or in the Terraform provider? If so, I'd prefer to set this to `null` here and use an upstream default.

I'd also prefer if we could keep the name equally to the upstream name here, so `master_ipv4_cidr_block` including the ipv4 part. I can only assume Google did this to be prepared for a potential ipv6 version of this attribute.

Spazzy757 replied:
This is required and there is no default, unfortunately, hence why I tried to use a range that hopefully nobody would use. also when it comes to terraform variables you cant use numbers in the name (hence why I removed the `ipv4` part)

pst replied:
We should split these two topics. Mixing in this one PR adds quite a bit of complexity. Especially given the upstream bug.

I suggest we leave this PR for private nodes only, and hardcode enable_private_endpoint to false. If required we can then add support for private endpoints in another PR where we can figure out what the right approach for the CIDR is. For what it's worth the [GKE docs use a 172 type address](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters). I have not looked into this enough to understand the implications of any default value.

Spazzy757 replied:
The problem is that the `master_ipv4_cidr_block ` is required when you set `master_ipv4_cidr_block ` to `true` not only `enable_private_endpoint`. I can look at updating it to the `172` range

Spazzy757 replied:
Just a note on this, I removed the `enable_private_endpoint ` as to my understanding this will put the API on a private network which will mean Terraform can not manage the resources within (which will mean Kubestacks manifests will not work without something akin to a bastion host)

pst replied:
No what I mean is, private nodes and private API are two separate features. Yes you may need to set both variables. But you don't have to expose it as variables outside the module. Simply make this PR only about private nodes so we can sort the bug out with upstream and find a way how to provide a migration path to users.

We leave private API for another PR later.

I hadn't read up on this before. But I did now and it seems to me, the way this works is it requires the new VPC-native cluster mode to not have public IPs for nodes. That's why `ip_allocation_policy` is required. Previously created clusters are routes-based. From what I found so far, there is no way to migrate a cluster from routes to VPC-nate without downtime.

pst replied:
It seems the default was changed for UI and CLI created clusters, but not for the Terraform module. I suppose this is due to no other backwards compatible option?

| In both gcloud and the Cloud Console, VPC-native is the default for new clusters and increasingly, GKE features such as Standalone Network Endpoint Groups (NEGs) have relied on clusters being VPC-native. In Terraform however, the default behaviour is to create a routes-based cluster for backwards compatibility.

Source: https://www.terraform.io/docs/providers/google/guides/using_gke_with_terraform.html#vpc-native-clusters

pst replied:
To move this forward I suggest you set default values that match what's in the Google docs but make the CIDRs configurable. Similar to https://github.com/kbst/terraform-kubestack/pull/135/commits/f6efd01f81fbc0a2c814e1a1f9055f6ae420ba78, so it's similar between the GKE and AKs modules. What do you think?

pst replied:
What do you think about the above @Spazzy757?

Spazzy757 replied:
Completely missed this, I can look at doing this tomorrow! I like it

pst commented on google/_modules/gke/cluster.tf:
```diff
@@ -49,4 +49,13 @@ resource "google_container_cluster" "current" {
+    master_ipv4_cidr_block  = var.master_cidr_block
+  }
+
+  ip_allocation_policy {}
```
```
Step #4 - "terraform plan":       + ip_allocation_policy { # forces replacement
Step #4 - "terraform plan":           + cluster_ipv4_cidr_block       = (known after apply)
Step #4 - "terraform plan":           + cluster_secondary_range_name  = (known after apply)
Step #4 - "terraform plan":           + node_ipv4_cidr_block          = (known after apply)
Step #4 - "terraform plan":           + services_ipv4_cidr_block      = (known after apply)
Step #4 - "terraform plan":           + services_secondary_range_name = (known after apply)
Step #4 - "terraform plan":           + subnetwork_name               = (known after apply)
Step #4 - "terraform plan":         }
```

pst replied:
Adding `ip_allocation_policy {}` forces a replacement of the cluster. So this update can not be done without downtime. Even with setting `enable_private_nodes = false` for backwards compatibility, if this attribute is included the cluster gets recreated. Need to find a solution for this.

Spazzy757 replied:
For internal networking, this is required as you need to enforce IP aliasing over routing (This is potentially why it would cause a recreate). Any ideas on how to get around it?

pst replied:
Try making this conditional like so:

```hcl
dynamic "ip_allocation_policy" {
  for_each = var.enable_private_nodes ? toset([1]) : null

  content {}
}
```
If `null` doesn't work in the example, try `toset([])`

This should get this blocker out of the way.

pst replied:
Also this, @Spazzy757?

Spazzy757 replied:
Again completely missed your last review will hack n this tomorrow

pst commented on google/_modules/gke/cluster.tf:
```diff
@@ -49,4 +49,13 @@ resource "google_container_cluster" "current" {
     }
   }
+
+  private_cluster_config {
```
```
Step #4 - "terraform plan":       + private_cluster_config {
Step #4 - "terraform plan":           + enable_private_nodes   = true # forces replacement
Step #4 - "terraform plan":           + master_ipv4_cidr_block = "10.254.0.0/28" # forces replacement
Step #4 - "terraform plan":           + peering_name           = (known after apply)
Step #4 - "terraform plan":           + private_endpoint       = (known after apply)
Step #4 - "terraform plan":           + public_endpoint        = (known after apply)
Step #4 - "terraform plan":         }
```

pst replied:
This is fine, especially since `enable_private_nodes = false` can be used to retain previous behavior. But setting this produces the following error currently:

```
Step #4 - "terraform plan": Error: Provider produced invalid plan
Step #4 - "terraform plan": 
Step #4 - "terraform plan": Provider "registry.terraform.io/-/google" has indicated "requires replacement"
Step #4 - "terraform plan": on module.gke_zero.module.cluster.google_container_cluster.current for a
Step #4 - "terraform plan": non-existent attribute path
Step #4 - "terraform plan": cty.Path{cty.GetAttrStep{Name:"private_cluster_config"},
Step #4 - "terraform plan": cty.IndexStep{Key:cty.NumberIntVal(0)},
Step #4 - "terraform plan": cty.GetAttrStep{Name:"master_ipv4_cidr_block"}}.
Step #4 - "terraform plan": 
Step #4 - "terraform plan": This is a bug in the provider, which should be reported in the provider's own
Step #4 - "terraform plan": issue tracker.
```

Need to report that upstream.

pst replied:
Upstream misunderstood this PR as trying to fix the issue and refused to look into it. Very unfortunate.

Spazzy757 replied:
Yeah I saw that, I thought they would reopen once you mentioned it

pst commented:
```
Step #4 - "terraform plan": Error: Provider produced invalid plan
Step #4 - "terraform plan": 
Step #4 - "terraform plan": Provider "registry.terraform.io/-/google" has indicated "requires replacement"
Step #4 - "terraform plan": on module.gke_zero.module.cluster.google_container_cluster.current for a
Step #4 - "terraform plan": non-existent attribute path
Step #4 - "terraform plan": cty.Path{cty.GetAttrStep{Name:"private_cluster_config"},
Step #4 - "terraform plan": cty.IndexStep{Key:cty.NumberIntVal(0)},
Step #4 - "terraform plan": cty.GetAttrStep{Name:"master_ipv4_cidr_block"}}.
Step #4 - "terraform plan": 
Step #4 - "terraform plan": This is a bug in the provider, which should be reported in the provider's own
Step #4 - "terraform plan": issue tracker.
```

Setting `enable_private_nodes = false` to keep previous behavior results in above error.

Spazzy757 commented:
This is going to take longer than I thought, there are a couple of issues im dealing with

pst commented:
I have yet to encounter a change that does not require dealing with multiple issues when it comes to Terraform, Kubernetes and the three cloud providers.

Spazzy757 commented:
@pst I have updated and it seems to work (I tested the changes and there are no changes if you set `enable_private_nodes = false` but if `enable_private_nodes = true` it enables them) let me know what you think, I used the Google Default for the master_cidr and removed the private api stuff and defaulted it to off

pst commented:
I tested this, it works as advertised. I like priavate nodes being the new default and I will document that existing clusters need to set a variable accordingly. Only nitpick is, I prefer the variable to be `disable_private_nodes = true` to disable and have it default to `false`. I think this is more natural, given that you'd only care about setting this var when you want to disable private nodes.

I'll take care of renaming the variable after the merge.

Thanks for your contribution @Spazzy757

pst commented:
Scrap that renaming comment. I oversaw that the upstream variable from the `google_container_cluster` is [`enable_private_nodes`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/container_cluster#enable_private_nodes). So it makes more sense to follow that, as you did.
